src/blockchain_db/lmdb/db_lmdb.cpp

```diff
@@ -3966,14 +3966,43 @@ std::vector<checkpoint_t> BlockchainLMDB::get_checkpoints_range(uint64_t start,
   return result;
 }
 
-uint64_t BlockchainLMDB::get_checkpoint_immutable_height() const
+bool BlockchainLMDB::get_immutable_checkpoint(checkpoint_t *immutable_checkpoint) const
 {
   size_t constexpr NUM_CHECKPOINTS = service_nodes::CHECKPOINT_NUM_CHECKPOINTS_FOR_CHAIN_FINALITY;
   std::vector<checkpoint_t> const checkpoints = get_checkpoints_range(height(), 0, NUM_CHECKPOINTS);
-  uint64_t result = 0;
-  if (checkpoints.size() == NUM_CHECKPOINTS)
-    result = checkpoints.back().height;
-  return result;
+
+  if (checkpoints.empty())
+    return false;
+
+  checkpoint_t const *immutable_checkpoint_ptr = nullptr;
+  if (checkpoints[0].type == checkpoint_type::service_node) // checkpoint[0] is the first closest checkpoint that is <= my height
+  {
+    // NOTE: The current checkpoint is a service node checkpoint. Go back
+    // 1 checkpoint, which will either be another service node checkpoint or
+    // a predefined one.
+    if (checkpoints.size() == 1)
+    {
+      // NOTE: Only one service node checkpoint recorded, we can override this checkpoint.
+      return false;
+    }
+    else
+    {
+      // If it's a service node checkpoint, this is the 2nd newest checkpoint,
+      // so we can't reorg past that height. If it's predefined, that's ok as
+      // well, we can't reorg past that height so irrespective, always accept
+      // the height of this next checkpoint.
+      immutable_checkpoint_ptr = &checkpoints[1];
+    }
+  }
+  else
+  {
+    immutable_checkpoint_ptr = &checkpoints[0];
+  }
```

Comment:
Slightly simpler logic (but this is subjective):  (The comments are great; just omitting them here for brevity).
```C++
checkpoint_t const *immutable_checkpoint_ptr = nullptr;
if (checkpoints[0].type != checkpoint_type::service_node)
    immutable_checkpoint_ptr = &checkpoints[0];
else if (checkpoints.size() > 1)
    immutable_checkpoint_ptr = &checkpoints[1];
else
    return false;

if (immutable_checkpoint) *immutable_checkpoint = *immutable_checkpoint_ptr;
```

Reply:
Another simplification: the pointer into the vector seems like unnecessary indirection versus just storing the value directly: (again omitting comments for brevity).

```C++
checkpoint_t immutable;
if (checkpoints[0].type != checkpoint_type::service_node)
    immutable = checkpoints[0];
else if (checkpoints.size() > 1)
    immutable = checkpoints[1];
else
    return false;

if (immutable_checkpoint) *immutable_checkpoint = immutable;

Reply:
Well I mean the non-ptr way is a copy right? You presuming this is optimised away in release? But also realised we can std::move to steal the signatures here as well.

Reply:
Oh, right, for some reason I was thinking it was just a `uint64_t`.